cpp/splinepy/proximity/proximity.hpp

```diff
@@ -0,0 +1,342 @@
+    grid_points_ = GridPoints_(parametric_bounds, resolutions);
+    coordinates_ = Coordinates_(
+        new typename SplineType::Coordinate_[grid_points_.Size()]
+    );
+
+    // lambda function to allow n-thread execution
+    auto sample_coordinates = [&] (int begin, int end) {
+      typename SplineType::ParametricCoordinate_ parametric_coordinate;
+      for (int i{begin}; i < end; ++i) {
+        grid_points_.IndexToParametricCoordinate(i, parametric_coordinate);
+        coordinates_[i] = spline_(parametric_coordinate);
+      }
+    };
+
+    // n-thread execution
+    splinepy::utils::NThreadExecution(sample_coordinates,
+                                      grid_points_.Size(),
+                                      n_thread);
+
+    // plant a new tree
+    coordinates_cloud_ = std::make_unique<Cloud_>(coordinates_,
+                                                  grid_points_.Size());
+    kdtree_ = std::make_unique<Tree_>(SplineType::kDim, *coordinates_cloud_);
+    kdtree_planted_ = true;
+  }
+
+  /// Make initial guess of choice
+  typename SplineType::ParametricCoordinate_ MakeInitialGuess(
+      InitialGuess initial_guess,
+      const double* goal) const {
+
+
+    if (initial_guess == InitialGuess::MidPoint) {
+      using ReturnValueType =
+          typename SplineType::ParametricCoordinate_::value_type;
+      const auto parametric_bounds =
+          splinepy::splines::GetParametricBounds(spline_);
+
+      // mid point is mean of parametric bounds. doesn't consider the goal.
+      return splinepy::utils::Mean<ReturnValueType>(
+          parametric_bounds[0],
+          parametric_bounds[1]
+      );
+
+    } else if (initial_guess == InitialGuess::KdTree) {
+      if (!kdtree_planted_) {
+        // hate to be aggresive, but here it is.
+        splinepy::utils::PrintAndThrowError(
+            "to use InitialGuess::Kdtree, please first plant the kdtree!"
+        );
+      }
+
+      // good to go. ask the tree
+      const int vote = 1;
+      int id[vote];
+      double distance[vote];
+      kdtree_->knnSearch(goal,
+                         vote /* closest neighbor */,
+                         &id[0],
+                         &distance[0]);
+
+      using ReturnType = typename SplineType::ParametricCoordinate_;
+
+      return 
+          grid_points_.template IndexToParametricCoordinate<ReturnType>(id[0]);
+    } else {
+      //well, shouldn't reach here, but to fight a warning, here it is
+      splinepy::utils::PrintAndThrowError("Invalid initial guess option!");
+      return typename SplineType::ParametricCoordinate_{};
+    }// if initial_guess == ...
+  }
+
+  /*!
+   * Builds RHS and fills "eyeders" matrix, which is also required in LHS.
+   *
+   * "eyeders" refer to collection of spline derivatives where only one
+   * parametric dimension is set to one and everythin else zero, which results
+   * in "eye"-like vectorized derivative query
```

Comment:
```suggestion
   * Calculates the first derivative of the distance 
   *
   * `spline_gradient` refers to the gradient of the spline position with respect 
   * to the individual parametric dimensions, whereas the rhs is the derivative of
   * the distance between the query points 
```
Please add formulas here

Reply:
`spline_gradient` sounds good